### PR TITLE
SYN-6716: `it:sec:cpe` issues when converting from 2.2->2.3 and 2.3->2.2

### PR DESCRIPTION
**Pull request by MichaelSquires**
bugfix: Updated CPE code to support better conversions from 2.2->2.3 and vice versa

### CODE REVIEW

**Review by invisig0th — APPROVED** [review]

File: synapse/tests/test_model_infotech.py [hunk 1]
```diff
@@ -27,7 +29,7 @@ async def test_infotech_basics(self):
             self.eq(nodes[0].get('url'), 'https://cwe.mitre.org/data/definitions/120.html')
             self.eq(nodes[0].get('parents'), ('CWE-119',))
 
-            self.eq(r'foo\:bar', core.model.type('it:sec:cpe').norm(r'cpe:2.3:a:foo\:bar:*:*:*:*:*:*:*:*:*')[1]['subs']['vendor'])
+            self.eq(r'foo:bar', core.model.type('it:sec:cpe').norm(r'cpe:2.3:a:foo\:bar:*:*:*:*:*:*:*:*:*')[1]['subs']['vendor'])
```
Comment (vEpiphyte) [inline synapse/tests/test_model_infotech.py:32]: Can we move this line  / concept to `test_infotech_cpes` ?

File: synapse/tests/test_model_inet.py [hunk 2]
```diff
@@ -2754,3 +2755,13 @@ async def test_model_inet_egress(self):
             self.eq(nodes[0].get('client'), 'tcp://1.2.3.4')
             self.eq(nodes[0].get('client:ipv4'), 0x01020304)
             self.eq(nodes[0].get('client:ipv6'), '::1')
+
+    async def test_cpe_scrape_one_to_one(self):
```
Comment (vEpiphyte) [inline synapse/tests/test_model_inet.py:2759]: I put this test in the wrong file :laughing:  lets move it to test_model_infotech

**Review by vEpiphyte — APPROVED** [review]

File: synapse/lib/modelrev.py [hunk 3]
```diff
@@ -993,6 +997,99 @@ async def save():
...
+                try:
+                    norm, info = form.type.norm(propvalu)
+                except s_exc.BadTypeValu as e: # pragma: no cover
+                    oldm = e.errinfo.get('mesg')
+                    logger.warning(f'Skipping {formname}={formvalu} : {oldm}')
+                    continue
+
+                if norm != formvalu:
+                    iden = s_common.ehex(buid)
+                    mesg = f'Unable to migrate primary property for {formname}="{formvalu}" (iden: {iden}). Continuing to migrate secondary properties.'
```
Comment (vEpiphyte) [inline synapse/lib/modelrev.py:1046]: This error message implies that we do migrate the primary property?

File: synapse/lib/modelrev.py [hunk 4]
```diff
@@ -993,6 +997,94 @@ async def save():
...
+
+            async def save():
+                await layr.storNodeEdits(nodeedits, meta)
+                nodeedits.clear()
+
+            genr = layr.liftByProp(form.name, propname)
+
+            async for _, buid, sode in genr:
+
+                sodevalu = sode.get('valu')
```
Comment (invisig0th) [inline synapse/lib/modelrev.py:1021]: This logic would preclude us normalizing a secondary property which was flying solo in the storage node. We should probably just let the prop check below stand.

**Review by invisig0th — CHANGES_REQUESTED** [review]

File: synapse/lib/modelrev.py [hunk 5]
```diff
@@ -993,6 +997,94 @@ async def save():
...
+                        continue
+
+                    try:
+                        subnorm, subinfo = subprop.type.norm(subvalu)
+                    except s_exc.BadTypeValu as e: # pragma: no cover
+                        oldm = e.errinfo.get('mesg')
+                        logger.warning(f'error norming subvalue {subprop.full}={subvalu}: {oldm}')
+                        continue
+
+                    subcurv = props.get(subprop.name)
```
Comment (vEpiphyte) [inline synapse/lib/modelrev.py:1065]: wont `subcurv` be `(valu, stortype)` ?

File: synapse/lib/modelrev.py [hunk 6]
```diff
@@ -993,6 +997,94 @@ async def save():
...
+
+                props = sode.get('props')
+                if props is None:
+                    continue  # pragma: no cover
+
+                propvalu = props.get(propname)
+                if propvalu is None: # pragma: no cover
+                    continue
+
+                propvalu = propvalu[0]
```
Comment (vEpiphyte) [inline synapse/lib/modelrev.py:1035]: this is slicing subcurv from (valu, stortype) right?

File: synapse/lib/modelrev.py [hunk 7]
```diff
@@ -993,6 +997,94 @@ async def save():
...
+                        continue
+
+                    subcurv = props.get(subprop.name)
+                    if subcurv is not None:
+                        if subcurv[1] != subprop.type.stortype: # pragma: no cover
+                            logger.warning(f'normFormSubs() may not be used to change storage types for {subprop.full}')
+                            continue
+                        subcurv = subcurv[0]
+
+                    if subcurv == subnorm:
```
Comment (vEpiphyte) [inline synapse/lib/modelrev.py:1072]: wont subcurv be (valu, stortype) and not match subnorm?

File: synapse/lib/modelrev.py [hunk 8]
```diff
@@ -993,6 +997,94 @@ async def save():
...
+                        subnorm, subinfo = subprop.type.norm(subvalu)
+                    except s_exc.BadTypeValu as e: # pragma: no cover
+                        oldm = e.errinfo.get('mesg')
+                        logger.warning(f'error norming subvalue {subprop.full}={subvalu}: {oldm}')
+                        continue
+
+                    subcurv = props.get(subprop.name)
+                    if subcurv is not None:
+                        if subcurv[1] != subprop.type.stortype: # pragma: no cover
+                            logger.warning(f'normFormSubs() may not be used to change storage types for {subprop.full}')
```
Comment (vEpiphyte) [inline synapse/lib/modelrev.py:1068]: wrong function name

File: synapse/lib/modelrev.py (hunk 5 above)
Comment (vEpiphyte) [inline synapse/lib/modelrev.py:1065]: We can do this check prior to the norm, right?

**Review by invisig0th — APPROVED** [review]

**Pull request merged**